### ISSUE TITLE
Add support for wrappers of callback-based APIs

### ISSUE THREAD

**Issue by frederikaalund** [agronholm/anyio#167]
First off, thanks for writing and maintaining anyio. It's a useful and well-written library.

I'm in the process of writing an anyio wrapper around paho-mqtt. Here is an example from the [paho-mqtt](https://github.com/eclipse/paho.mqtt.python) readme:

```python
# The callback for when a PUBLISH message is received from the server.
def on_message(client, userdata, msg):
    print(msg.topic+" "+str(msg.payload))

client = mqtt.Client()
client.on_message = on_message
```

To bridge the gap between the old-school callbacks and the async domain, I use `create_memory_object_stream`:

```python
send_stream, receive_stream = create_memory_object_stream()

def _send_to_stream(client, userdata, msg):
    try:
        send_stream.send_nowait(msg)  # This is an async function so I can't call it in this context
    except WouldBlock:
        ...

client.on_message = _send_to_stream
```

This approach doesn't work, however, since [`send_nowait` is an async function](https://github.com/agronholm/anyio/blob/0455cd9b2b0e08ce916476e22e4f198c45a15f83/src/anyio/streams/memory.py#L121). From a cursory look, I get the impression that curio support is the culprit. Because `Event.set` is async in curio, `send_nowait` is also async.

Is there some way around this? Can I somehow specify that I don't need curio support and get the non-async version of `Event.set` and, in turn, the non-async version of `send_nowait`?

If there currently isn't such functionality, would you be interested in a pull request that adds this?

Many other old-school APIs are callback-based. [RxPy](https://github.com/ReactiveX/RxPY) is another example. I'd also like to write anyio helpers for RxPy to bridge the gap to the async domain. Currently, my efforts are blocked by the above-mentioned restriction.

This issue relates to #99.

**Comment by agronholm** [comment]
Are you suggesting that I add an API to AnyIO that only works on some of the backends?

**Comment by frederikaalund** [comment]
Thanks for the quick reply.

> Are you suggesting that I add an API to AnyIO that only works on some of the backends?

Not if there is a way around it.

I recently read about [`curio.UniversalEvent`](https://github.com/dabeaz/curio/blob/dc11ffefb2101b1b37be4c2183b5b3beac604fca/curio/sync.py#L56). This class has an non-async version of `Event.set`. Can we replace `Event` with `UniversalEvent` in the curio backend? In turn, this opens up for non-async versions of `send_nowait`.

**Comment by smurfix** [comment]
I would suggest that this is not the right approach. Callbacks lose most of the nice Structured Concurrency restrictions that "force" us to write correct async code.

Stupid question, why are you browbeating paho into the async world when there already is at least one working anyio-compatible MQTT client+server? (distmqtt)

While Rx is based on callbacks, that's not a viable async structure. In an async world I'd start with implementing its `.subscribe` method like this:
```
def no_op(*_): pass
async def maybe_async(p, *a):
    res = p(*a)
    if iscoroutine(res):
        res = await res
    return res

async def subscribe(on_next: no_op, on_error: no_op, on_completed: no_op):
    async with self.subscription() as sub:
        async for result in sub:
            try:
                item = result.unwrap()  # assuming it's an Outcome
            except Exception as exc:
                await maybe_async(on_error, exc)
            else:
                await maybe_async(on_next, item)
        await maybe_async(on_completed) 
```
and thus start with rewriting Rx's data sources as async iterators. No callbacks required. If the user wants them (admittedly it's easier for the end user to write one-liners that way), OK, but its internals shouldn't use them and its API shouldn't require them.

**Comment by frederikaalund** [comment]
Thanks for the feedback, smurfix.

> Stupid question, why are you browbeating paho into the async world when there already is at least one working anyio-compatible MQTT client+server? (distmqtt)

Good question! There's lot of maturity in the time-proven, callback-based APIs. For some projects, I prefer this maturity over the newer APIs from the async world. I want to create simple async wrappers around the callback-based APIs to bridge the gap and get the best of both worlds.

Such bridges already exists for:
 * [trio-paho-mqtt](https://github.com/bkanuka/trio-paho-mqtt)
 * [asyncio-mqtt](https://github.com/sbtinstruments/asyncio-mqtt) (this one is maintained by me)

I want to create a new bridge with anyio (maybe call it anyio-mqtt?).

Aside: I think that async-first libraries such as hbmqtt and distmqtt may very will become the new standard with time.

> I would suggest that this is not the right approach. Callbacks lose most of the nice Structured Concurrency restrictions that "force" us to write correct async code.

I totally agree. Callbacks are not the way to go. A proper bridge hides all the underlying callbacks and only exposes an idiomatic async interface. Seems to work for asyncio-mqtt and trio-paho-mqtt. You mention some restrictions that may break this illusion. Care to elaborate? :)

> and thus start with rewriting Rx's data sources as async iterators. No callbacks required. If the user wants them (admittedly it's easier for the end user to write one-liners that way), OK, but its internals shouldn't use them and its API shouldn't require them.

Indeed, this is a cleaner approach. I don't want to rewrite RxPy and maintain that rewrite, however. I prefer a simply bridge to rely on the existing maturity of RxPy. Maybe I'm foolish. :)

**Comment by smurfix** [comment]
> There's lot of maturity in the time-proven, callback-based APIs.

Well, IMHO it's a local maximum – that breaks down as soon as you embed it in larger systems. Which is why people invented Structured Concurrency in the first place.

> You mention some restrictions that may break this illusion. Care to elaborate? :)

Internally you often can't get around callbacks or futures. An async network client protocol (like MQTT) for example must multiplex requests and dispatch the incoming responses to some set of possibly-unrelated tasks.

The idea is to encapsulate that structural violation in such a way that it's no longer visible in your API.

**Comment by frederikaalund** [comment]
> Internally you often can't get around callbacks or futures. An async network client protocol (like MQTT) for example must multiplex requests and dispatch the incoming responses to some set of possibly-unrelated tasks.
> 
> The idea is to encapsulate that structural violation in such a way that it's no longer visible in your API.

Totally agree.

In essence, I want to implement `anyio-mqtt` in the following stack:

    network API --> paho-mqtt --> callback-based API --> anyio-mqtt --> idiomatic async API

Here, `anyio-mqtt` hides the underlying callback-based API. Problem is that for this to work, I need a primitive in `anyio` that let's me communicate from sync code (i.e., the callbacks) to async code.

Currently, the curio backend is blocking this sync-to-async communication. I think that we can use `curio.UniversalEvent` to get around this restriction. I'd be happy to make a pull request for this if you agree with the approach. I can even implement this as a non-breaking addition to the existing API. E.g., `Event.set_sync` and `MemoryObjectSendStream.set_nowait_sync`.

**Comment by agronholm** [comment]
I've recently run into the same problem with [asyncpg](https://github.com/MagicStack/asyncpg). The only way to listen to notifications is with a synchronous callback. But since the library itself was only compatible with asyncio, I decided to go with an asyncio Event to catch the notifications.

What's your plan? Making an AnyIO UniversalEvent, or making all events of Curio UniversalEvents?

**Comment by frederikaalund** [comment]
Thanks for the prompt replies. I'm glad to see that there is some shared interest on this topic.

> What's your plan? Making an AnyIO UniversalEvent, or making all events on Curio into UniversalEvents?

I'm open to suggestions. I had not considered an `anyio.UniversalEvent` until now.

---

Pros of an `anyio.UniversalEvent`:
 * No changes to the existing `anyio.Event` API.

Cons of an `anyio.UniversalEvent`:
 * Maybe it's a but confusing for the user to have both `UniversalEvent` and `Event`. After all, both trio and asyncio only have a single event class (with the semantics of `UniversalEvent`)

Note that curio also has `curio.UniversalQueue` which is the analogue to `asyncio.Queue` and `trio.Channel`. If we add `anyio.UniversalEvent` should we then also add `anyio.UniversalObjectStream` to be consistent?

---

Pros of replacing `curio.Event` with `curio.UniversalEvent` in the curio back-end:
 * There is only a single `anyio.Event` class. This is similar to the asyncio and trio APIs.

Cons of replacing `curio.Event` with `curio.UniversalEvent` in the curio back-end:
 * To avoid breaking the existing API, we add an `anyio.Event.set_sync` function. This bloats the API.
 * Similarly, we add a `_sync` variant to the other functions that currently use `anio.Event.set` internally. E.g.: `MemoryObjectSendStream.set_nowait_sync`. More API bloat.

In a future major release, we deprecate the async variants and replace them with the `_sync` version.

Is there a reason why `anyio.Event.set` is currently async besides for the sake of the curio backend? Maybe I'm missing some use case.

---

Either approach will work for me. What do you prefer?

**Comment by smurfix** [comment]
Well, curio implements the multi-call-ability of its `UniversalEvent.set()` by inspecting the caller. That's … not something I'd recommend. In fact its not a solution I'd contemplate using, let alone implementing, in the first place.

I'd add `anyio.create_sync_event` which tells anyio to use the "Universal" version of `Event`, adding a `*_sync` method. On curio that'd use UniversalEvent instead of Event. (I wouldn't use UniversalEvent everywhere.)

Also we should tell people that this will probably go away again, sometime in the future, and long-term they should please ask their library providers to replace these callback mechanisms with something structured. Adding a callback mechanism on top of that, for compatibility, is trivial after all.

**Comment by frederikaalund** [comment]
> Well, curio implements the multi-call-ability of its UniversalEvent.set() by inspecting the caller. That's … not something I'd recommend. In fact its not a solution I'd contemplate using, let alone implementing, in the first place.

I agree. Seems rather hacky. :)

> I'd add anyio.create_sync_event which tells anyio to use the "Universal" version of Event, adding a *_sync method. On curio that'd use UniversalEvent instead of Event. (I wouldn't use UniversalEvent everywhere.)

That works for me.

> Also we should tell people that this will probably go away again, sometime in the future, [...]

These "Universal" semantics are part of both trio's and asyncio's APIs (in both their `Event` and `Queue`/`Channel` classes). Maybe it's a good idea to not force an async API when it's not needed. That is, why implement `Event.set` as async if it can easily be sync? Or am I misinformed here? Are there actually a significant "cost" associated with these combined sync-async ("Universal") semantics?

> [...] and long-term they should please ask their library providers to replace these callback mechanisms with something structured. Adding a callback mechanism on top of that, for compatibility, is trivial after all.

I agree.

**Comment by smurfix** [comment]
Calling an async (as opposed to sync) version of some function that itself doesn't call the scheduler has little enough overhead not to matter.

Apparently curio does call the scheduler when you set the event asynchronously, possibly on the grounds that the code waiting for it should be allowed to run*now* instead of whenever the setter gets around to do await things next. This might cut down on latency a little bit but I'd rather have a choice, and if I do need it then `anyio.sleep(0)` is easy enough to type in the next line.

**Comment by frederikaalund** [comment]
Sorry, I was a bit unclear in my wording. I was more thinking along the lines of: Why impose additional restrictions on `Event.set` by making it `async` when the sync version works just fine.

I can call a sync function from *both* a sync context and an async context.
I can *only* call an async function from an async context.

So when we choose to make `Event.set` async, we are imposing restrictions. Do we really want to do this?

**Comment by agronholm** [comment]
> Sorry, I was a bit unclear in my wording. I was more thinking along the lines of: Why impose additional restrictions on Event.set by making it async when the sync version works just fine.

Was your criticism aimed towards a proposed new API, the existing `Event` interface in AnyIO or the Curio Event class?

**Comment by smurfix** [comment]
@frederikaalund  The problem is/was that `UniversalEvent` is a fairly new addition to curio AFAIK and when anyio's API was invented there was no sync version. (That and I wouldn't recommend using `UniversalEvent` if you don't **really** need it).

So now we're stuck with adding workarounds.

**Comment by frederikaalund** [comment]
> @frederikaalund The problem is/was that UniversalEvent is a fairly new addition to curio AFAIK and when anyio's API was invented there was no sync version.

Ah, that figures. That explains why `anyio.Event.set` is currently async. :)

> Was your criticism aimed towards a proposed new API, the existing Event interface in AnyIO or the Curio Event class?

I'm just trying to understand what you prefer for the proposed new API. :) Both short-term and long-term.

My take is that *short-term* we simply add `_sync` variants to the existing `anyio` methods. E.g.: `anyio.Event.set_sync`. I think we are all more-or-less agreed on that.

For the *long-term*, I'm more conflicted. Will we simply deprecate these `_sync` methods in hope that other library APIs implement a proper async interface (and not only callbacks)? Or, will we replace the existing async methods with their sync counterparts?

I lean towards the latter option but that is just my personal opinion. :) I base my opinion on the existing asyncio and trio APIs where `asyncio.Event.set` and `trio.Event.set` are sync already. Similarly, I see that curio (recently) got `curio.UniversalEvent.set` that is also sync. Now that all of the three backends have a sync version, why would we choose to make `anyio.Event.set` async in the long-term?

I ask this question not because I'm fundamentally opposed to making it async but just because I don't understand the reasoning. I'll make the pull request in any case, I'd just like to understand your preferences on the long-term. :)

**Comment by agronholm** [comment]
> Ah, that figures. That explains why anyio.Event.set is currently async. :)

No, it doesn't really. `UniversalEvent` is meant specifically for situations where you need to communicate across the sync/async barrier. It's too heavyweight for general purpose use.

> I ask this question not because I'm fundamentally opposed to making it async but just because I don't understand the reasoning. I'll make the pull request in any case, I'd just like to understand your preferences on the long-term. :)

This all stems from David Beazley's philosophy where everything that works with the event loop should be marked `async`, so that you wouldn't accidentally try to use it from synchronous code. This philosophy has caused me plenty of headache and I've almost dropped curio support entirely because of it. But it wouldn't be "any I/O" if it only supported two async libraries, would it?

Solving this problem will probably require some new feature in AnyIO because I cannot think of a good way to implement this with the primitives currently offered by its API.

**Comment by agronholm** [comment]
Would you actually even be able to use this with backends other than asyncio anyway?

**Comment by frederikaalund** [comment]
Thanks for clarifying. It really helps me better understand the API design decisions.

> Would you actually even be able to use this with backends other than asyncio anyway?

I had both trio and asyncio in mind when I initially proposed this. I know that it's possible to idiomatically wrap a callback-based API with either library since I've seen it done. I didn't give curio much thought before this discussion. I guess we can make it work in curio too with `curio.UniversalEvent`. I don't know if `curio.UniversalEvent` is too heavyweight for practical use, though. Never used it.

**Comment by agronholm** [comment]
My point is that such callback based APIs tend to be centered on a specific async library. What do we gain by adding such a mechanism? Your original intent was to build an AnyIO wrapper around paho-mqtt, right? What is the intended end result then, if paho-mqtt only supports asyncio? How did you plan to add support for the rest of the backends, or did you?

**Comment by frederikaalund** [comment]
First off, thanks for all of the feedback on this issue. It's really appreciated. :) That goes to both of you.

> What is the intended end result then, if paho-mqtt only supports asyncio? How did you plan to add support for the rest of the backends, or did you?

I've been vague in my wording on this. Sorry about that. Let me try again. paho-mqtt is not tied to a specific event loop. paho-mqtt comes with some built-in options but also [provides a callback-based API through which we can attach an external event loop](https://github.com/eclipse/paho.mqtt.python#external-event-loop-support).

Here is a quote from paho-mqtt's README:

> There are four options for managing the network loop. Three are described here, the fourth in "External event loop support" below. Do not mix the different loop functions.

Soo paho-mqtt is actually event loop agnostic. That's why I think that paho-mqtt is a perfect fit for an anyio wrapper (`anyio-mqtt`):

    paho-mqtt <--> external event loop API <--> anyio-mqtt <--> idiomatic async API

I hope that clears things up. Let me know if you have other questions or comments.

**Comment by smurfix** [comment]
Well … you could use the `greenback` module to call back into async land.

**Comment by frederikaalund** [comment]
> Well … you could use the greenback module to call back into async land.

Never heard about greenback before. Thanks for sharing. :)

From a cursory look, I think that greenback will technically work for this. However, I'm concerned about the practical aspect of pulling in the greenlet library to solve this. Seems a bit overkill to have two async frameworks in a single application. Especially when the purpose of one of them is to "fix" the sync/async communication of the other one.

I'd much rather see direct support for async/sync communication in anyio itself.

**Comment by agronholm** [comment]
This has been fixed in master.

**Comment by frederikaalund** [comment]
Thank you very much. :)
